[ArteTV] config api v2

casaper opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [ ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Switch to ArteTV API v2, per #29640:
* resolves #29640
* resolves #29870
* resolves #30878
* resolves #30919.

Also many closed duplicate issues:
#30912
#30927
#30930
#30951
#30963
#30980
#31006
#31020
#31242
#31292
#31389
#31455
#31510

casaper committed:
WIP: arte config api v2

dirkf committed:
Update from yt-dlp

dirkf committed:
Merge branch 'master' into extractor/arte/improvement-switch-to-config-api-v2

dirkf commented:
I hope you don't mind. I've updated this from a working WIP extractor bac-ported from yt-dlp.

dirkf commented on youtube_dl/extractor/arte.py:
```diff
@@ -275,27 +323,23 @@ def suitable(cls, url):
+
+        return merge_dicts(
+            self.playlist_from_matches(items, playlist_id=playlist_id, playlist_title=title),
+            {'description': self._og_search_description(webpage, default=None)})
```
```suggestion
            {'description': self._og_search_description(webpage, default=None)})

```

dirkf committed:
Linted?

casaper commented:
> I hope you don't mind. I've updated this from a working WIP extractor bac-ported from yt-dlp.

Not at all, go ahead. I forgot this even exists.

Thx for finishing it.

dirkf commented:
Great, thanks.

I see that somehow the yt-dlp code didn't need to get the Bearer token, which makes things easier.

dirkf committed:
Update

Based on yt-dlp/yt-dlp 051d6b4 merging [extractor/arte] yt-dlp/yt-dlp#3302: Move to v2 API

Authored by: fstirlitz, pukkandan

dirkf commented on youtube_dl/extractor/arte.py:
```diff
@@ -4,185 +4,232 @@
+        mobj = self._match_valid_url(url)
         video_id = mobj.group('id')
         lang = mobj.group('lang') or mobj.group('lang_2')
+        langauge_code = self._LANG_MAP.get(lang)
+
+        config = self._download_json('{0}/config/{1}/{2}'.format(self._API_BASE, lang, video_id), video_id)
+
+        geoblocking = traverse_obj(config, ('data', 'attributes', 'restriction', 'geoblocking')) or {}
+        if geoblocking.get('restrictedArea'):
+            raise GeoRestrictedError('Video restricted to {0!r}'.format(geoblocking['code']),
+                                     countries=self._COUNTRIES_MAP.get(geoblocking['code'], ('DE', 'FR')))
+
+        if not traverse_obj(config, ('data', 'attributes', 'rights')):
+            # Eg: https://www.arte.tv/de/videos/097407-215-A/28-minuten
+            # Eg: https://www.arte.tv/es/videos/104351-002-A/serviteur-du-peuple-1-23
+            raise ExtractorError(
+                'Video is not available in this language edition of Arte or broadcast rights expired', expected=True)
+
+        formats, subtitles = [], {}
+        secondary_formats = []
+        for stream in config['data']['attributes']['streams']:
+            # official player contains code like `e.get("versions")[0].eStat.ml5`
+            stream_version = stream['versions'][0]
+            stream_version_code = stream_version['eStat']['ml5']
+
+            lang_pref = -1
+            m = self._VERSION_CODE_RE.match(stream_version_code)
+            if m:
+                lang_pref = int(''.join('01'[x] for x in (
+                    m.group('vlang') == langauge_code,      # we prefer voice in the requested language
+                    not m.group('audio_desc'),              # and not the audio description version
+                    bool(m.group('original_voice')),        # but if voice is not in the requested language, at least choose the original voice
+                    m.group('sub_lang') == langauge_code,   # if subtitles are present, we prefer them in the requested language
```
```suggestion
        language_code = self._LANG_MAP.get(lang)

        config = self._download_json('{0}/config/{1}/{2}'.format(self._API_BASE, lang, video_id), video_id)

        geoblocking = traverse_obj(config, ('data', 'attributes', 'restriction', 'geoblocking')) or {}
        if geoblocking.get('restrictedArea'):
            raise GeoRestrictedError('Video restricted to {0!r}'.format(geoblocking['code']),
                                     countries=self._COUNTRIES_MAP.get(geoblocking['code'], ('DE', 'FR')))

        if not traverse_obj(config, ('data', 'attributes', 'rights')):
            # Eg: https://www.arte.tv/de/videos/097407-215-A/28-minuten
            # Eg: https://www.arte.tv/es/videos/104351-002-A/serviteur-du-peuple-1-23
            raise ExtractorError(
                'Video is not available in this language edition of Arte or broadcast rights expired', expected=True)

        formats, subtitles = [], {}
        secondary_formats = []
        for stream in config['data']['attributes']['streams']:
            # official player contains code like `e.get("versions")[0].eStat.ml5`
            stream_version = stream['versions'][0]
            stream_version_code = stream_version['eStat']['ml5']

            lang_pref = -1
            m = self._VERSION_CODE_RE.match(stream_version_code)
            if m:
                lang_pref = int(''.join('01'[x] for x in (
                    m.group('vlang') == language_code,      # we prefer voice in the requested language
                    not m.group('audio_desc'),              # and not the audio description version
                    bool(m.group('original_voice')),        # but if voice is not in the requested language, at least choose the original voice
                    m.group('sub_lang') == language_code,   # if subtitles are present, we prefer them in the requested language
```

dirkf committed:
`langauge_code`

dirkf committed:
Fix ArteTVEmbedIE

dirkf commented on youtube_dl/extractor/arte.py:
```diff
@@ -4,208 +4,272 @@
+
+    # yt-dlp shims
+
+    @classmethod
+    def _match_valid_url(cls, url):
+        return re.match(cls._VALID_URL, url)
+
```
Already ported.
```suggestion
```

dirkf commented on youtube_dl/extractor/arte.py:
```diff
@@ -4,208 +4,272 @@
+                )))
+
+            short_label = traverse_obj(stream_version, 'shortLabel', expected_type=compat_str, default='?')
+            if stream['protocol'].startswith('HLS'):
```
```suggestion
            if stream['protocol'].startswith(('HLS', 'API_HLS')):
```

dirkf reviewed (commented):
Fix https://github.com/ytdl-org/youtube-dl/issues/30919#issuecomment-4752444297
